macGist/Keychain.swift

```diff
@@ -84,15 +123,124 @@ struct Keychain {
         return password
     }
     
-    func deleteItem() throws {
-        // Delete the existing item from the keychain.
-        let query = keychainQuery(withService: KeychainConfiguration.serviceName, account: KeychainConfiguration.account, accessGroup: KeychainConfiguration.accessGroup)
+    func deleteToken() throws {
+        let query = KeychainConfiguration.query()
+        let status = SecItemDelete(query as CFDictionary)
+        
+        guard status == noErr || status == errSecItemNotFound else { throw KeychainError.unhandledError(status: status) }
+    }
+    
+    func deleteGitHubClientId() throws {
+        let query = KeychainConfiguration.GitHub.Gist.Account.clientId.query()
+        let status = SecItemDelete(query as CFDictionary)
+        
+        guard status == noErr || status == errSecItemNotFound else { throw KeychainError.unhandledError(status: status) }
+    }
+    
+    func deleteGitHubClientSecret() throws {
+        let query = KeychainConfiguration.GitHub.Gist.Account.clientSecret.query()
         let status = SecItemDelete(query as CFDictionary)
         
-        // Throw an error if an unexpected status was returned.
         guard status == noErr || status == errSecItemNotFound else { throw KeychainError.unhandledError(status: status) }
     }
     
+    func save(gitHubClientId: String) throws {
+        let encodedGitHubClientId = gitHubClientId.data(using: String.Encoding.utf8)!
+        
+        do {
+            try _ = readGitHubClientId()
+            
+            var attributesToUpdate = [String : AnyObject]()
+            attributesToUpdate[kSecValueData as String] = encodedGitHubClientId as AnyObject?
+            
+            let query = KeychainConfiguration.GitHub.Gist.Account.clientId.query()
+            let status = SecItemUpdate(query as CFDictionary, attributesToUpdate as CFDictionary)
+            
+            guard status == noErr else { throw KeychainError.unhandledError(status: status) }
+        } catch KeychainError.noGitHubClientId {
+            var newItem = KeychainConfiguration.GitHub.Gist.Account.clientId.query()
+            newItem[kSecValueData as String] = encodedGitHubClientId as AnyObject?
+            
+            let status = SecItemAdd(newItem as CFDictionary, nil)
+            
+            guard status == noErr else { throw KeychainError.unhandledError(status: status) }
+        }
+    }
+    
+    func readGitHubClientId() throws -> String {
+        var query = KeychainConfiguration.GitHub.Gist.Account.clientId.query()
+        
+        query[kSecMatchLimit as String] = kSecMatchLimitOne
```

Comment:
to be honest, when I see `as String` a lot of times, I feel not comfortable... 😐

Reply:
Removed a lot of casts. Created a typealias for `[CFString:AnyObject]` so now the query will be of this type.